File: completed_python/베스트앨범.py
from collections import defaultdict
from operator import itemgetter

def solution(genres, plays):
    count_list = defaultdict(int)
    total_list = defaultdict(list)

    for song_id, genre, play in zip(counter(), genres, plays):
        count_list[genre] += play
        total_list[genre].append((-play, song_id))

    genres_in_order = sorted(count_list.keys(), key=lambda g:count_list[g], reverse=True)
    print(count_list.keys())

    best_songs = []

    for genre in genres_in_order:
        sorted_list = sorted(total_list[genre])
        temp = []
        for _, song_id in sorted_list:
            temp.append(song_id)
        best_songs.extend(temp[:2])

    return best_songs


def solution2(genres, plays):
    play_count_by_genre = defaultdict(int)
    songs_in_genre = defaultdict(list)

    for song_id, genre, play in zip(counter(), genres, plays):
        play_count_by_genre[genre] += play
        songs_in_genre[genre].append((-play, song_id))

    # print(songs_in_genre)

    genre_in_order = sorted(play_count_by_genre.keys(), key=lambda g:play_count_by_genre[g], reverse=True)

    answer = []

    for genre in genre_in_order:
        answer.extend([ song_id for minus_play, song_id in sorted(songs_in_genre[genre])][:2])

    return answer


# def solution3(generes, plays):
#     total = []
#     for song_id, genre, play in zip(counter(), genres, plays):
#         total.append([song_id, genre, play])
#     print(sorted(total, key=itemgetter()))


def counter():
     i = 0
     while True:
         yield i
         i += 1


genres = ["classic", "pop", "classic", "classic", "pop"]
plays = [500, 600, 150, 800, 2500]
    
# print(counter())
# print(solution(genres, plays))
print(solution2(genres, plays))
# solution3(genres, plays)

File: completed_php/베스트앨범.php
<?php 

$genres = ["classic", "pop", "classic", "classic", "pop"];
$plays = [500, 600, 150, 800, 2500];

// 1. 속한 노래의 총 카운트가 높은 장르부터 수록
$counter = [];

foreach($genres as $key => $value):
    if(!array_key_exists($value, $counter)){
        $counter[$value] = $plays[$key];
    } else {
        $counter[$value] += $plays[$key];
    }
endforeach;

arsort($counter); // 2. 높은 장르가 무엇인지 확인

// 3. 높은 장르부터 추출하기 위해 두 배열이 합산된 토탈 배열을 생성

$total = [];

foreach($genres as $key => $value):
    $total[$value][] = [$plays[$key], $key];
endforeach;

$result = [];

// 4. 카운팅 수가 높은 배열부터 2개씩만 돌아가며 ID를 추출
foreach($counter as $genre => $count):
    arsort($total[$genre]);
    foreach(array_slice($total[$genre], 0, 2) as $temp):
        $result[] = $temp[1];
    endforeach;
endforeach;

var_dump($result);